src/core/facebook-auth.js

```diff
@@ -38,19 +38,21 @@ XMPPSocialProvider.prototype.login = function(loginOpts, continuation) {
       return this.oauth.launchAuthFlow(url, stateObj);
     }.bind(this)).then(function(continuation, responseUrl) {
       var token = responseUrl.match(/access_token=([^&]+)/)[1];
-      var xhr = new XMLHttpRequest();
+      var xhr = freedom["core.xhr"]();
       xhr.open('GET', 'https://graph.facebook.com/me?access_token='+token);
-      xhr.onload = function(continuation, token, xhr) {
-        var response = JSON.parse(xhr.responseText);
-        var credentials = {
-          jid: '-'+response.id+'@chat.facebook.com',
-          access_token: token,
-          api_key: this.oAuthAppSecret,  // secret, not id!
-          host: 'chat.facebook.com'
-        };
-        this.logger.log('Got facebook credentials: ' + JSON.stringify(credentials));
-        this.onCredentials(continuation, {cmd: 'auth', message: credentials});
-      }.bind(this, continuation, token, xhr);
+      xhr.on("onload", function(continuation, token, xhr) {
+        xhr.getResponseText().then(function(continuation, token, responseText) {
+          var response = JSON.parse(responseText);
+          var credentials = {
+            jid: '-'+response.id+'@chat.facebook.com',
+            access_token: token,
+            api_key: this.oAuthAppSecret,  // secret, not id!
+            host: 'chat.facebook.com'
+          };
+          this.logger.log('Got facebook credentials: ' + JSON.stringify(credentials));
+          this.onCredentials(continuation, {cmd: 'auth', message: credentials});
+        }.bind(this, continuation, token));
+      }.bind(this, continuation, token, xhr));
```

Comment:
Maybe just paranoia. The concern here is that when login gets called, it will need to await user actions (to login). If we happen to call social.login again and initiate a new auth flow (e.g. user presses login button again because they accidentally closed their tab, we don't want to confuse which continuation to call or which token to use.